### PR TITLE
fix: Make window resize commands manage cmdheight

### PR DESCRIPTION
**Pull request by famiu**
Previously, the window resize commands did not resize the value of `cmdheight` when they caused a change in the topframe height, leaving a gap between the end of topframe and the start of the command line, this PR fixes that by making window resize commands automatically change the value of cmdheight if the resize affects the height of topframe.

Not sure why the test in `test/legacy/cmdline_spec.lua` is failing. Would really like some help in fixing it, since I failed to fix it by myself despite trying really hard.

### CODE REVIEW

File: src/nvim/window.c [hunk 1]
```diff
@@ -5222,10 +5222,16 @@ void win_setheight_win(int height, win_T *win)
     if (full_screen && msg_scrolled == 0 && row < cmdline_row) {
       grid_fill(&default_grid, row, cmdline_row, 0, Columns, ' ', ' ', 0);
     }
+    if (msg_grid.chars) {
```
Comment (bfredl) [inline src/nvim/window.c:5225]: wouldn't this be nested inside the previous `if` (the space was increased) ?

Comment (famiu) [reply src/nvim/window.c]: ah, good catch, fixed it

**Comment by famiu** [comment]
@bfredl you think you might have any idea why the test in `legacy/cmdline_spec.lua` is failing? I can't for the life of me figure it out

**Comment by famiu** [comment]
> > honestly not sure what could be causing it if it's in enter_tabpage() though
> 
> The test is about switching tabs and seeing if that clears cmdline or not rofl
> 

I know that, I'm saying I don't know what in `enter_tabpage()` could cause the screen to *not* be cleared

**Comment by zeertzjq** [comment]
Note: there are now `win_move_statusline()` and `win_move_separator()` that manages `'cmdheight'` correctly, I think.

**Comment by famiu** [comment]
> Note: there are now `win_move_statusline()` and `win_move_separator()` that manage `'cmdheight'` correctly, I think.

That doesn't fix the bugs with resizing topframe though? Correct me if I'm wrong

**Comment by zeertzjq** [comment]
You can ignore that comment.

File: src/nvim/window.c [hunk 2]
```diff
@@ -5221,11 +5221,17 @@ void win_setheight_win(int height, win_T *win)
...
+        clear_cmdline = true;
+      }
     }
     cmdline_row = row;
+    p_ch = MAX(Rows - cmdline_row, 1);
+    curtab->tp_ch_used = p_ch;
     msg_row = row;
     msg_col = 0;
     redraw_all_later(NOT_VALID);
+    showmode();
```
Comment (bfredl) [inline src/nvim/window.c:5234]: would setting `redraw_cmdline = true` also work? Then cmdline will be drawn as part of the entire screen

Comment (famiu) [reply src/nvim/window.c:5234]: Test fails even if I do that

Comment (bfredl) [reply src/nvim/window.c:5234]: hmm you mean the code is broken, but this change makes no difference?

Comment (famiu) [reply src/nvim/window.c:5234]: The test could be broken as well honestly, but yeah, the change makes no difference

File: src/nvim/window.c [hunk 3]
```diff
@@ -5452,7 +5458,9 @@ static void frame_setheight(frame_T *curfrp, int height)
   if (curfrp->fr_parent == NULL) {
     // topframe: can only change the command line
     if (height > ROWS_AVAIL) {
-      height = ROWS_AVAIL;
+      // If height is greater than the available space, try to create space for the frame by
+      // reducing 'cmdheight' if possible, while making sure `cmdheight` doesn't go below 1.
+      height = MIN(ROWS_AVAIL + (p_ch - 1), height);
```
Comment (mityu) [inline src/nvim/window.c:5463]: This PR is to automatically change the `'cmdheight'` value when the command line height is changed with `:resize` like operations.
However, it seems to me that these lines are for another purpose: reduce the command line height when maximizing a window (or making window height bigger).
This change is a breaking change and I wonder if it is expected.

Comment (zeertzjq) [reply src/nvim/window.c:5463]: When `'cmdheight'` is greater than `1` this also causes to `wincmd _` to change `'cmdheight'` to `1`.

Comment (mityu) [reply src/nvim/window.c:5463]: @zeertzjq  Thank you for completing my comment!

Here's some screenshots to show what behavior is changed.
All the binaries I used are downloaded from the GitHub Releases.
- v0.6.1
cmdheight-v061

- v0.7.0
cmdheight-v070

**Comment by Shougo** [comment]
@famiu Could you response @mityu 's question?

**Comment by famiu** [comment]
> @famiu Could you response @mityu 's question?

I don't see any questions here. The change @mityu mentioned was intentionally done by me, not sure if it counts as a breaking change. @zeertzjq would probably be able to answer that, though.

**Pull request merged**